File: js/src/protocol.ts
/**
 * Protocol constants for EHBP (Encrypted HTTP Body Protocol)
 */
export const PROTOCOL = {
  ENCAPSULATED_KEY_HEADER: 'Ehbp-Encapsulated-Key',
  RESPONSE_NONCE_HEADER: 'Ehbp-Response-Nonce',
  KEYS_MEDIA_TYPE: 'application/ohttp-keys',
  KEYS_PATH: '/.well-known/hpke-keys',
  PROBLEM_JSON_MEDIA_TYPE: 'application/problem+json',
  KEY_CONFIG_PROBLEM_TYPE: 'urn:ietf:params:ehbp:error:key-config',
} as const;

/**
 * HPKE suite configuration matching the Go implementation
 */
export const HPKE_CONFIG = {
  KEM: 0x0020, // X25519 HKDF SHA256
  KDF: 0x0001, // HKDF SHA256
  AEAD: 0x0002 // AES-256-GCM
} as const;

/**
 * EHBP-WS constants (SPEC Section 8): encrypted WebSocket channels
 * secured with the Noise NK handshake, keyed by the server's X25519
 * identity key.
 */
export const NOISE_WS = {
  /** WebSocket subprotocol negotiated during the upgrade. */
  SUBPROTOCOL: 'ehbp.noise.v1',
  /** Noise protocol name; also the first input to the handshake hash. */
  PROTOCOL_NAME: 'Noise_NK_25519_AESGCM_SHA256',
  /** Prologue bound into the handshake hash. */
  PROLOGUE: 'ehbp noise websocket v1',
  /** Record type carrying application data. */
  RECORD_DATA: 0x01,
  /** Record type signaling authenticated channel termination. */
  RECORD_CLOSE: 0x02,
  /** Default cap on a record's plaintext payload (1 MiB). */
  DEFAULT_MAX_MESSAGE_SIZE: 1 << 20,
  /** Read-limit allowance for the record type byte and AEAD tag. */
  RECORD_OVERHEAD: 64,
  /** Cap on handshake message size. */
  HANDSHAKE_READ_LIMIT: 4096,
  /** Default cap on the dial-plus-handshake duration (SPEC Section 8). */
  HANDSHAKE_TIMEOUT_MS: 10_000,
  /** Records per direction before the sending key is ratcheted. */
  REKEY_INTERVAL: 1 << 16,
} as const;

File: rust/src/protocol.rs
pub const ENCAPSULATED_KEY_HEADER: &str = "Ehbp-Encapsulated-Key";
pub const RESPONSE_NONCE_HEADER: &str = "Ehbp-Response-Nonce";
pub const KEYS_MEDIA_TYPE: &str = "application/ohttp-keys";
pub const KEYS_PATH: &str = "/.well-known/hpke-keys";
pub const PROBLEM_JSON_MEDIA_TYPE: &str = "application/problem+json";
pub const KEY_CONFIG_PROBLEM_TYPE: &str = "urn:ietf:params:ehbp:error:key-config";

pub const KEY_ID: u8 = 0;
pub const KEM_X25519_HKDF_SHA256: u16 = 0x0020;
pub const KDF_HKDF_SHA256: u16 = 0x0001;
pub const AEAD_AES_256_GCM: u16 = 0x0002;

pub const HPKE_REQUEST_INFO: &[u8] = b"ehbp request";
pub const EXPORT_LABEL: &[u8] = b"ehbp response";
pub const EXPORT_LENGTH: usize = 32;
pub const REQUEST_ENC_LENGTH: usize = 32;
pub const RESPONSE_NONCE_LENGTH: usize = 32;
pub const AES256_KEY_LENGTH: usize = 32;
pub const AES_GCM_NONCE_LENGTH: usize = 12;
pub const RESPONSE_KEY_LABEL: &[u8] = b"key";
pub const RESPONSE_NONCE_LABEL: &[u8] = b"nonce";

pub const WS_SUBPROTOCOL: &str = "ehbp.noise.v1";
pub const NOISE_PROTOCOL_NAME: &str = "Noise_NK_25519_AESGCM_SHA256";
pub const NOISE_PROLOGUE: &[u8] = b"ehbp noise websocket v1";
pub const WS_RECORD_DATA: u8 = 0x01;
pub const WS_RECORD_CLOSE: u8 = 0x02;
pub const DEFAULT_WS_MAX_MESSAGE_SIZE: usize = 1 << 20;
/// Framing added to a record payload: 1 record type byte, a 16-byte AEAD
/// tag, and margin for WebSocket read limit accounting.
pub const WS_RECORD_OVERHEAD: usize = 64;
pub const WS_HANDSHAKE_READ_LIMIT: usize = 4096;
pub const WS_REKEY_INTERVAL: u64 = 1 << 16;
/// Default cap on the WebSocket dial plus Noise handshake (SPEC Section 8.4).
pub const WS_HANDSHAKE_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
